### PR TITLE
#33: add note taking tools

### PR DESCRIPTION
**Pull request by perrygoy**
Here's the best i could come up with for this note-taking endeavor. I think i might have gotten sidetracked and things got very complicated. I don't like where things currently stand. I'm pushing up this PR to get some eyes on it. 

There are currently a lot of `mypy` errors with the `NotedMixin` that i cannot resolve, and i don't like just how *weird* it is. 

I think if i just went with my first instinct of creating more resolutions like `IsEqualToTheNoted("key name")`, it might be better. I was just worried that there were going to be a lot more resolutions added. Seeing now how many were changed to mixins, i think it wouldn't be so bad!

### CODE REVIEW

File: tests/test_resolutions.py [hunk 1]
```diff
@@ -20,15 +20,28 @@
 )
 
 
-class TestBaseAbility:
```
Comment (perrygoy) [inline tests/test_resolutions.py:23]: Copy/paste error on this test name, finally fixed!

**Comment by perrygoy** [comment]
I've changed this PR to not include any changes to resolutions. Using `TakeNotes` in a Resolution just wan't working, so why try to make it work? 

I think being able to take notes is still valuable, there will just need to be an extra step where the actor checks their notes before they get to the assertion section. I think this is OK. Now things will look like this:

```python
the_actor.attempts_to(
    Visit(URL),
    MakeNote.of(TheText.of_the(WELCOME_MESSAGE)).as_("first message"),
    LogIn(),
    Wait.for_the(WELCOME_MESSAGE).to_appear(),
)

first_message = the_actor.uses_ability_to(TakeNotes).to_look_up("first message")

the_actor.should_see_the(
    (Text.of_the(WELCOME_MESSAGE), DoesNot(ReadExactly(first_message))),
)
```

**Comment by WandyYing** [comment]
Hi @perrygoy Good Job. It's close to perfection. I think it would be better if it could be changed to the following one.
```
the_actor.attempts_to(
    Visit(URL),
    MakeNote.of(TheText.of_the(WELCOME_MESSAGE)).as_("first message"),
    LogIn(),
    Wait.for_the(WELCOME_MESSAGE).to_appear(),
)

# first_message = the_actor.uses_ability_to(TakeNotes).to_look_up("first message") # delete it

the_actor.should_see_the(
    (Text.of_the(WELCOME_MESSAGE), DoesNot(ReadExactly(Note.of(first_message)))),
)
```

How do you like add action `Note.of` to retrieves the previously recorded `first_message`?

**Comment by perrygoy** [comment]
I totally agree, that is the ideal. Unfortunately, the `should_see_the` method of the `Actor` is expecting a tuple of a Auestion and a Resolution, it isn't expecting to need to figure out an Action in the middle of that. That was the issue we kept running into while discussing on #33 and while i was creating this PR.

The other issue is that we're using a specific `Actor`'s ability to `TakeNotes`, so that `Note.of` action would need to know what actor to ask for the note. It's all a big mess!

I did think of something while responding to your message though (a `Director` class...???). It's probably not going to lead anywhere, so don't get your hopes up. I think this PR is the best i can make it, for now.

**Comment by WandyYing** [comment]
>The other issue is that we're using a specific Actor's ability to TakeNotes, so that Note.of action would need to know what actor to ask for the note. It's all a big mess!

Yes, It's all a big mess.  And so far, I also think this PR is the best way. 

About `Note.of` action, its implementation may need to change many places, or even the lower level design. Things can't be perfect. We can shelve them for future evolution.

**Comment by perrygoy** [comment]
@WandyYing hoo boy check out what that `Director` comment turned into!

**Comment by WandyYing** [comment]
@perrygoy So far, I think it good for me. gj :)

File: docs/topics/director.rst [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Using the Director
+------------------
+
+The director is always present,
+but you may not need them
+in every test.
+All you need to do
+is call them,
+like ``Director()``,
+to summon them.
```
Comment (pgoy-dod) [inline docs/topics/director.rst:20]: It seems to me that this is not the envisioned "typical" use of the `Director`, so maybe move this further down as a tip instead. Most folks won't be using the `Director` directly, right?

**Review by pgoy-dod — APPROVED** [review]
While this PR got a bit chunky, i'm very excited about the new `Director` class. I imagine that class will come in handy for a lot of things!

I also like how *svelte* the Resolutions are now. Much easier to write new ones. My approval comes with the caveat that you will fix the small issue in my comment, @perrygoy.

**Comment by WandyYing** [comment]
@perrygoy `looks_up` needs to capture the exception that the key does not exist.

**Comment by perrygoy** [comment]
> @perrygoy `looks_up` needs to capture the exception that the key does not exist.

While it could do that, i think it's important that it doesn't. I don't have any other information to add to that exception (like in the cases where i catch `WebDriverException` elsewhere to add the Target), so bubbling up the raw `KeyError`—i think—is the right thing to do.

Is there an advantage to catching the exception in `looks_up`?

**Comment by WandyYing** [comment]
When a user uses a wrong key, a humanized exception description can make it easier for him to understand. For example, `The look up "{key}" has never been recorded. balabala` is better then `KeyError: key`

Because I don't think we can assume that all users are Python programmers. There are still junior or business people to read use cases and report.

**Pull request merged**